### PR TITLE
users: Retry account unlocking if once isn't enough

### LINKED ISSUE #1216: Account locked via password doesn't unlock consistently via Cockpit
An account locked via passwd doesn't unlock consistently via Cockpit, sometimes requiring the action twice.

 1. login to cockpit as root and open "Account" card of existing user
 2. in command line do "passwd -l <username>" as root
 3. go back to Cockpit and unlock the user
 4. wait a few seconds

### PR DESCRIPTION
**Pull request by dperpeet**
There are different mechanisms for account locking/unlocking.
Sometimes an unlock operation will succeed but leave the
user account in a locked state [0]. If Cockpit tries to change the
locked state of an account and the operation succeeds but the
actual locked state remains unchanged, try one more time.

[0] https://bugzilla.redhat.com/show_bug.cgi?id=1142234

Fixes #1216

### CODE REVIEW

File: pkg/users/local.js [hunk 1]
```diff
@@ -733,12 +733,23 @@ PageAccount.prototype = {
            });
     },
 
-    get_locked: function() {
+    get_locked: function(desired_lock_state) {
```
Comment (stefwalter) [inline pkg/users/local.js]: It seems backwards to be using a function called ```get_locked()``` to change state and execute commands. Shouldn't the desired_lock_state logic be elsewhere?

**Comment by dperpeet** [comment]
I changed the `get_locked` function to return a promise and moved the verification into `change_locked`, where it belongs.

File: pkg/users/local.js [hunk 2]
```diff
@@ -734,18 +734,26 @@ PageAccount.prototype = {
            });
     },
 
-    get_locked: function() {
+    get_locked: function(update_display) {
+        update_display = typeof update_display !== 'undefined' ? update_display : true;
```
Comment (stefwalter) [inline pkg/users/local.js]: What does this line do? Seems like it can be removed.

File: pkg/users/local.js [hunk 3]
```diff
@@ -734,18 +734,26 @@ PageAccount.prototype = {
...
                       { "environ": [ "LC_ALL=C" ], "superuser": "require" })
-           .done(parse_locked)
-           .fail(log_unexpected_error);
+            .done(function(content) {
+                    self.locked = parse_locked(content);
+                    if (update_display)
+                        self.update();
+                    dfd.resolve(self.locked);
+                })
+            .fail(dfd.reject);
```
Comment (stefwalter) [inline pkg/users/local.js]: This invokes reject with the wrong 'this'

File: pkg/users/local.js [hunk 4]
```diff
@@ -930,11 +938,29 @@ PageAccount.prototype = {
...
                        this.account["name"],
-                       $('#account-locked').prop('checked') ? "--lock" : "--unlock"], { "superuser": "require"})
-           .done($.proxy (this, "get_locked"))
+                       desired_lock_state ? "--lock" : "--unlock"], { "superuser": "require"})
+            .done(function() {
+                self.get_locked(false)
+                    .done(function(locked) {
+                        /* if we care about what the lock state should be and it doesn't match, try to change again
+                           this is a workaround for different ways of handling a locked account
+                           https://github.com/cockpit-project/cockpit/issues/1216
```
Comment (stefwalter) [inline pkg/users/local.js]: Is there a bug about this in bugzilla?

File: pkg/users/local.js (hunk 2 above)
Comment (dperpeet) [inline pkg/users/local.js]: This makes the default action of get_locked to call update - which happens in the code.
Only in the specific case that we need to try setting the value again do we care about setting this to false, that's why there's a default value.

File: pkg/users/local.js (hunk 4 above)
Comment (dperpeet) [inline pkg/users/local.js]: added reference

File: pkg/users/local.js (hunk 3 above)
Comment (dperpeet) [inline pkg/users/local.js]: this should be fixed now

**Pull request closed without merging**